8.x 1.x

mccrodp opened the pull request:
Here is my work so far on D8.

Separated the tests out and converted them to be D8 compatible. There are a few unknowns so far, such as the entity_token dependency, which I presume is no longer needed, so it is commented out with a @todo marker.

The tests look like a good start to me, please take a look when you get a chance. They run, but cause all cause a SchemaIncompleteException, probably as they cannot successfully enable the message and message UI modules. I'll come back to these when I convert the module code itself.

mccrodp commented:
**So far:**
- Converted all _hook_menu_ entries to routes, including local tasks and contextual links yaml files
- Added a dynamic route entry with associated _MessageRoutes_ class for devel route entries
- Added almost all entries pointing to a controller method as we will perform access control there
- Added Forms at src/Form 

**Todo / Open Issues:**
- Forms and the _MessageUiController_ need a lot of work still, next on my list
- Unsure what the _message_ui_test_ module does really. Will we need it as a module in D8?
- Did you already convert _message_notify_ui_? I seem to recall you mentioning something about this

RoySegall commented on message_ui.links.contextual.yml:
```diff
@@ -0,0 +1,11 @@
+  route_name: message_ui.delete_message
+  group: message-ui
+  title: 'Delete'
+  weight: 10
```
New line in the end of the file.

RoySegall commented on message_ui.links.task.yml:
```diff
@@ -0,0 +1,18 @@
+  route_name: message_ui.delete_message
+  base_route: message_ui.show_message
+  title: 'Delete'
+  weight: 10
```
New line in the end of the file.

RoySegall commented on message_ui.routing.yml:
```diff
@@ -0,0 +1,43 @@
+    _controller: '\Drupal\message_ui\Controller\MessageUiController::delete_multiple'
+
+route_callbacks:
+  - '\Drupal\message_ui\Routing\ExampleRoutes::routes'
```
New line in the end of the file.

RoySegall commented:
First of all 10x for the work. Tag me in any new things in the future because i didn't see this one.

Did you try the module convertor? it can convert menu items and test to Drupal 8. Could be useful.

mccrodp commented:
@RoySegall thanks for the feedback. 

I had heard of the Drupal Module Upgrader, but as this is my first D8 upgrade I wanted to take a more hands on approach. However, on closer inspection there is an analyze command `dmu-analyze`. I will use this and also compare output of the `dmu-upgrade` command with what I have already, see if I missed anything. 

I'm going to look at this again before next week hopefully and in the meantime I added those "New line in the end of the file" changes to the yaml files and committed the changes.

mccrodp commented:
I have temporarily added, during the upgrade process, the [html output from running the drupalmoduleupgrader analyze command on the latest version of message_ui](http://opensourceopenmind.ninja/sites/default/files/tmp/upgrade-info.html) on 7.x branch as a reference.

mccrodp commented:
@RoySegall - I've had a small bit more time for this recently, so have committed some changes. I will have more time for this again soon and also throughout Drupalcon Barcelona I'll work on this whenever I get time.

I've updated `entity_load()` and `user_load()` calls in _message_ui_ and all `l()` and `user_access()` calls. I will leave the _message_notify_ui_ submodule until afterwards. Did you say you had converted this elsewhere already? I'm not sure if I got an answer about that.

There is an [open issue on drupalmoduleupgrader for running the dmu-upgrade](https://www.drupal.org/node/2555203). I'll compare what I have from manual conversion to the auto conversion to D8 when that is resolved.

RoySegall commented:
Iv'e ported message to to Drupal 8. I'll try to take a look on what youve done so far.

mccrodp commented:
Yep, I'm referring to your Message for D8 as I go along. I should have a more substantial amount of work to review soon. I've only done about 25% or so of what is needed until I can step through the module code usefully with XDebug.

mccrodp commented:
A few items I'm currently unsure of now. I will go into more detail later, but just to flag the biggest unknowns at the moment.
- The use of the QueueWorker and associated methods (at the moment I have added some methods to QueueWorker as static methods which I don't think is quite right).
- The best place for the access control checks and associated methods, currently in MessageAccessControlHandler.php

mccrodp commented:
Hi @RoySegall. I've gotten this quite a bit further now, but I'm running into some more structural unknowns along with the above previously mentioned. For example, I have a central MessageUIController where I redirect my routes to, however it might be better to route directly to newly created Action classes (such as your \Drupal\node\Plugin\Action\DeleteMessage class) or to call my EntityFormController classes directly? https://www.drupal.org/node/1734556

If you get a chance to have even a quick scan of the structure of my route yaml considering the above and let me know which is the preferred approach re Forms, Access, Actions, etc. I would be grateful.

In the meantime I will keep working on it and perhaps make further progress on _message_notify_ui_. Thanks.

RoySegall commented:
Can you give a commit access so i could do any changes. I'll set up the things and see how it goes.

mccrodp commented:
Added you as a collaborator to my fork of the project and opened up the Issue tracking on the fork also incase it's useful for separating issues / comments. Thanks.

RoySegall commented:
10x, i'm looking for now. I'll try to comment today or tomorrow.

RoySegall commented:
I'm looking on the work but the basic work flow is not working. I can't access the creation page of a message even if i'm admin.

mccrodp commented:
Hi, yes I know it's not working at all yet. Could you please look at the components in code and let me know what I need to reorganise structurally (from points mentioned in : https://github.com/RoySegall/message_ui/pull/2#issuecomment-139504286) before I can start testing in the UI. I've merely just reorganised to a more D8 structure and converted existing code, but it's not quite right. Once I'm redirected on any major adjustments structurally I'll start going through the UI and get the workflow on track and working correctly. My next step will be stepping through the UI, starting with the message creation page, but any overall comments on the above mentioned in the meantime is much appreciated. Thanks.

RoySegall commented:
I can't tell if the code is OK or not before the module do something. Start apply the basic work flow and we can the code later.

mccrodp commented:
No problem, yep, D8 all very new and unknown at the moment :) 

You're right, was just wondering if anything stood out as completely wrong from my work so far. I'll get the basic workflow in place and after that I'll ping you again. Thanks!

RoySegall commented:
As far as i can see there is not best practice for D8 so we kind off going in the dark for now. Like wise for OG8, we have kind of a strategy.

mccrodp commented:
True, get it working first and improve towards best practices later, and as you say: it'll be easier to review once there is something working / partially working.

...upgrading OG to D8, that will be an interesting / challenging task! ;)

RoySegall commented:
Yes, the best minds are on it!

Tag me when you have something or catch me in the near by DrupalCon

mccrodp commented:
Great, good luck with it!

Yes, hopefully catch you at Drupalcon and if I get a chance to get a good portion of the workflow working in the meantime I'll tag you again.

mccrodp commented:
Hey @RoySegall - Have a lot more of the workflow in place now and (almost) working.

Main items in front of me to solve now:
- Get <code>MessageViewController</code> or <code>message/{message}</code> route to call _message.html.twig_ correctly when viewing a message; at the moment only the _region.html.twig_ is called
- Get <code>MessageForm</code> to display _message_type_ fields correctly and populate with _message_ values

mccrodp commented:
Continuing [message_ui port on a separate branch](https://github.com/mccrodp/message_ui/tree/update-d8-stable) until https://github.com/Gizra/message/pull/37 gets accepted into Message module. Made quite a bit of progress. FYI - the reason the fields weren't showing on the form when we were looking at Drupalcon was that they were set to "hidden" in the new "Manage Form Display" UI, haha. Will update on more concrete progress soon.

mccrodp commented:
@RoySegall - on the above linked branch in https://github.com/RoySegall/message_ui/pull/2#issuecomment-159637555 I have gotten the local tasks working now and have at least the following remaining:
- Ensure entity is properly saved on form Save, displayed on "View" tab and edit form fields are populated correctly on message edit
- Access Control on Message entities similar to that of Node, but simplified
- Port Queue worker functionality for arguments (may need advice on this one)

I'll merge the _message_ui_ work to date into this pull request once https://github.com/Gizra/message/pull/37 or https://github.com/Gizra/message/pull/39 are merged.

mccrodp commented:
Merged back to main PR branch. Changes made in conjunction with Message module with the following PRs merged together [https://github.com/Gizra/message/pull/39, https://github.com/Gizra/message/pull/48, https://github.com/Gizra/message/pull/50, https://github.com/Gizra/message/pull/51].

_Items remaining:_
1. Get tests running
2. Ensure Access Control is being fired / working
3. Research and fix `MessageUiArgumentsWorker` QueueWorker class
4. Ensure all test are green
5. Release dev release following release of Message

mccrodp commented:
Good progress on 1. - Tests are now "running" to an extent and travis build file seems to be doing the job: https://travis-ci.org/mccrodp/message_ui/jobs/103951386. Just need to solve the views schema issues in yml file on Message module https://github.com/Gizra/message/issues/52 and get the tests running to completion, even if failing.

mccrodp commented:
Tests now running due to changes in PR to fix message schema: https://github.com/Gizra/message/pull/53

https://travis-ci.org/mccrodp/message_ui/jobs/104319376

mccrodp commented:
@RoySegall - Getting much closer with this, not so much left to do:

1 Clean up and correct Access methods
2 Fix Arguments worker

I think I should be ok with 1, but will need to check how the arguments worker functions in D7. Any ideas on this / No. 2 for D8?

At the moment the _.travis.yml_ file points to my branch in this PR (https://github.com/Gizra/message/pull/57) as without valid schema, the tests cannot run / initialise Message module.

Failing due to 1 & 2 above, but running to completion and a lot of green lights on everything else: https://travis-ci.org/mccrodp/message_ui/builds/109325401

**EDIT**: Should those new Message module "owner" getters and setters be in the `MessageInterface` also?

mccrodp commented:
Hey @RoySegall, any idea what this Exception may mean: 

`'Drupal\Component\Plugin\Exception\PluginNotFoundException: "The "entity:message:foo" plugin does not exist."`

This error is encountered after a call to `drupalPostForm()` to save a message entity  in tests.": https://travis-ci.org/mccrodp/message_ui/jobs/121126540

I tried adding a _SaveMessage_ plugin similar to the plugin in node module, but no luck. https://github.com/mccrodp/message_ui/blob/8.x-1.x/src/Plugin/Action/SaveMessage.php

Tried asking a few times on IRC, but I'm a bit stuck with it. Any help appreciated!!

mccrodp commented:
Also, what's the best way to merge this PR? It's quite big, should we have smaller PRs, or will you just merge when it gets closer to completion as there is no 8.x branch to merge to at the moment. Thanks @RoySegall

mccrodp commented:
I opened issues for the remaining tasks: https://github.com/mccrodp/message_ui/issues

mccrodp commented:
@RoySegall - Some _message_ui_ goals for this week: https://github.com/mccrodp/message_ui/wiki/Development-Wiki

RoySegall commented on .travis.yml:
```diff
@@ -0,0 +1,92 @@
+  # - composer self-update
+
+  # Drush.
+  - composer global require "youngj/httpserver:dev-master#41dd2b7"
```
We might want to use 8.0.5 version.

mccrodp replied:
Done.

RoySegall replied:
Can't see the changes over this one.

mccrodp replied:
See line 20 & 63:

https://github.com/mccrodp/message_ui/blob/8.x-1.x/.travis.yml#L20
https://github.com/mccrodp/message_ui/blob/8.x-1.x/.travis.yml#L63

RoySegall commented on .travis.yml:
```diff
@@ -0,0 +1,92 @@
+  # as well on fatal errors.
+  - TEST_OUTPUT=$(! egrep -i "([0-9]+ fails)|(PHP Fatal error)|([0-9]+ exceptions)" /tmp/test.txt > /dev/null)$?; echo $TEST_OUTPUT
+  # Exit the build
+  - if [ $TEST_EXIT -eq 0 ] && [ $TEST_OUTPUT -eq 0 ]; then exit 0; else exit 1; fi
```
missing line in the end of the file

mccrodp replied:
Done

RoySegall commented on config/install/message_ui.settings.yml:
```diff
@@ -0,0 +1,6 @@
+  how_to_act: 'update_when_added'
+  number_items: 250
+show_preview: true
+anonymous: 'Anonymous' # What is this variable and how best to store/access it?
```
If we are not using setting any where so let's remove it.

mccrodp replied:
That is a comment I added at the start, I'll remove the comment. We use this setting at: `/src/Form/MessageForm.php:57`

RoySegall commented on config/install/system.action.message_save_action.yml:
```diff
@@ -0,0 +1,10 @@
+label: 'Save message'
+type: message
+plugin: message_save_action
+configuration: {  }
```
Missing line in the end of the file.

RoySegall commented on message_notify_ui/message_notify_ui.info.yml:
```diff
@@ -0,0 +1,12 @@
+
+dependencies:
+ - views
+ - views_bulk_operations
```
I don't think the vbo module live as it is.

RoySegall commented on message_notify_ui/src/Plugin/Views/field/NotifyButton.php:
```diff
@@ -0,0 +1,48 @@
+    if (user_access('send message notify')) {
+      return l(t('Notify'), 'message/' . $message->id() . '/notify');
+    }
+    return NULL;
```
Just return. No need for NULL.

RoySegall commented on message_ui.js:
```diff
@@ -1,9 +1,9 @@
-(function ($) {
+(function ($, Drupal, drupalSettings) {
```
Let's move it back to the js folder.

RoySegall commented on message_ui.libraries.yml:
```diff
@@ -0,0 +1,9 @@
+    - core/jquery
+    - core/drupal
+    - core/drupalSettings
+    - core/drupal.form
```
Missing line in the end of the file.

RoySegall commented on message_ui.links.action.yml:
```diff
@@ -0,0 +1,6 @@
+  route_name: message_ui.add_page
+  title: 'Add message'
+  appears_on:
+    - view.message.page_1
```
Missing line in the end of the file.

RoySegall commented on message_ui.links.contextual.yml:
```diff
@@ -0,0 +1,10 @@
+  route_name: entity.message.delete_form
+  group: message-ui
+  title: 'Delete'
+  weight: 10
```
Missing line in the end of the file.

RoySegall commented on message_ui.links.task.yml:
```diff
@@ -0,0 +1,21 @@
+  route_name:  entity.message.delete_form
+  base_route:  entity.message.canonical
+  title: Delete
+  weight: 10
```
Missing line in the end of the file.

RoySegall commented on message_ui.module:
```diff
@@ -1,447 +1,245 @@
-        'access callback' => 'message_ui_access_control',
-      );
+function message_ui_entity_base_field_info_alter(&$fields, EntityTypeInterface $entity_type) {
+    // Alter the uid and created field to include display settings.
```
what the use of this one?

mccrodp replied:
@RoySegall It allows the "Authoring Information" or "Owner Information" tab in the vertical tabs to show on the entity form. In node this is in Node.php, but we either have to put this in Message.php or here in an alter hook.

RoySegall replied:
OK. It seems we have indentation issues here.

RoySegall commented on message_ui.module:
```diff
@@ -1,447 +1,245 @@
+function message_ui_entity_type_alter(array &$entity_types) {
+    if (isset($entity_types['message'])) {
+
+        /**
```
The indentation looks wrong.

If this one commented out then remove it.

RoySegall commented on message_ui.module:
```diff
@@ -1,447 +1,245 @@
-    ),
-  );
+    return array(
+        'message_add_list' => array(
```
Indentation looks wrong.

RoySegall commented on message_ui.module:
```diff
@@ -1,447 +1,245 @@
-  $output .= '</ul>';
-
-  return $output;
+function template_preprocess_message_add_list(&$variables) {
```
Indentation looks wrong here as well.

RoySegall commented on message_ui.module:
```diff
@@ -1,447 +1,245 @@
-  $items = array();
-  $allowed_types = message_ui_user_can_create_message();
+function message_ui_form_message_user_admin_settings_alter(&$form, \Drupal\Core\Form\FormStateInterface $form_state) {
+    $form['update_tokens'] = array(
```
Indentation looks wrong here as well.

RoySegall commented on message_ui.module:
```diff
@@ -454,443 +252,92 @@ function message_ui_form_message_user_admin_settings_alter(&$form, $form_state)
-    $token_name = str_replace(array('@{', '}'), array('[', ']'), $token);
-    $value = token_replace($token_name, array('message' => $message));
+    $query = \Drupal::entityQuery('message');
+    $result = $query
```
Indentation

RoySegall commented on message_ui_test/message_ui_test.info.yml:
```diff
@@ -0,0 +1,10 @@
+dependencies:
+ - message
+ - views
+ - message_ui
```
Missing line in the end of the file.

RoySegall commented on src/Controller/MessageController.php:
```diff
@@ -0,0 +1,106 @@
+    $build['#content'] = $content;
+
+    if (!empty($build)) {
+      return $build;
```
You can remove the `$build` array and do some thing like:

``` php
return [
  ...
];
```

And not declare the variable.

RoySegall commented on src/Entity/MessageRouteProvider.php:
```diff
@@ -0,0 +1,43 @@
+
+    return $route_collection;
+  }
+}
```
Missing line in the end of the file.

RoySegall commented on src/Form/DeleteMultiple.php:
```diff
@@ -0,0 +1,189 @@
+
+    $form['actions']['cancel']['#href'] = $this->getCancelRoute();
+
+    // @todo - from message_ui in D7.
```
If commented out then remove it.

RoySegall commented on src/QueueWorker/MessageArgumentsWorker.php:
```diff
@@ -0,0 +1,124 @@
+  }
+}
+
+?>
```
No need for the close tag and there is a missing line in the end of the file.

RoySegall commented on src/Routing/MessageRoutes.php:
```diff
@@ -0,0 +1,66 @@
+  public function routes() {
+    $routes = array();
+
+    // @todo Complete code for interaction with devel module for this menu item.
```
if it's commented out let's remove it.

RoySegall commented on src/Tests/MessageUiHardCodedArguments.php:
```diff
@@ -0,0 +1,111 @@
+    );
+    $this->drupalPostForm('message/' . $message->id() . '/edit', $edit, t('Update'));
+    $this->assertText('Dummy name', 'The hard coded token was updated with a custom value.');
+  }
```
An empty line is needed between the last method and the closure of the message.

RoySegall commented on src/Tests/MessageUiMassiveHardCodedArguments.php:
```diff
@@ -0,0 +1,117 @@
+
+  /**
+   * The user object.
+   * @var
```
Add a line and specify the type of the property. it's probably `AccountInterface` or something similar.

mccrodp commented:
Hi @RoySegall - I have added the step to show fields to the README. I have made the remaining code review changes also.

Re: showing message text, I was using the default class `EntityViewBuilder` so I changed it to `MessageViewBuilder`, however I cannot see the partials being displayed. The build array is generated by the Message module, it includes the partials. I am not sure if this working.

Can we merge, then I will look at this in the Message module and make a PR on message or alter on Message UI, whichever is appropriate separately to this. Be nice to get close this big ugly PR :) Thanks!

RoySegall commented on message_ui.module:
```diff
@@ -1,447 +1,226 @@
-        );
-      }
+function message_ui_entity_type_alter(array &$entity_types) {
+    if (isset($entity_types['message'])) {
```
Indentation issues here as well.

RoySegall commented on message_ui.module:
```diff
@@ -454,443 +233,92 @@ function message_ui_form_message_user_admin_settings_alter(&$form, $form_state)
-function message_ui_instance_message_manage_validate($form, &$form_state) {
-  field_attach_form_validate('message', $form_state['#entity'], $form, $form_state);
-}
+        // Get all the messages.
```
Indentation issues over here.

RoySegall commented on src/Entity/MessageRouteProvider.php:
```diff
@@ -0,0 +1,43 @@
+    $route_collection = new RouteCollection();
+
+    $route = (new Route('/message/{message}/edit'))
+        ->setDefault('_entity_form', 'message.edit')
```
Indentation issues over here

RoySegall commented:
@mccrodp It seems that you have indentation issues. Please check they are set to 2 spaced in you IDE.

mccrodp commented:
@RoySegall - done, I committed reformatted code. I had set it in the IDE, but had to go through each file and reformat individually :)

RoySegall commented on config/install/message_ui.settings.yml:
```diff
@@ -0,0 +1,6 @@
+  how_to_act: 'update_when_added'
+  number_items: 250
+show_preview: true
+anonymous: 'Anonymous'
```
Missing line in the end of the file.

RoySegall commented:
I'll merge the branch once I'll apply the small fixes.

mccrodp commented:
@RoySegall - added the new line and remove a comment from travis, so the `composer install` is line 62 now.

RoySegall commented:
Giving this another look there is no messages to create:
![add_message___site-install_and_develeventsubscriber_php_-_drupal8_-___applications_mamp_htdocs_drupal8_](https://cloud.githubusercontent.com/assets/1222368/16543986/9563a27c-40fb-11e6-9c37-26c33953d414.jpg)

mccrodp commented:
@RoySegall Thank you. I had forgotten to add the `#theme` to the build array since I had removed that local `$build` variable. Fixed now.

RoySegall commented:
I testes it again and when I created a message comment i got a fata error.

mccrodp commented:
Really? What's the error? I'm using it no problem with Message Private module's private_message type at moment. I do have an open issue on Message module's MessageViewBuilder at the moment though, was it that? Also a PR on Message I am using with this module, might be related.

I'll take a look with comment message type soon and see. Some day we will get this merged :)

mccrodp commented:
Hi @RoySegall I didn't look at this exactly as the work on renaming Message Type to Message Template was merged.

I renamed all to "message template" and I was able to add, edit, delete a message via message_ui.

[I couldn't check with message_example due: https://github.com/Gizra/message/issues/114]

mccrodp commented:
Hi @RoySegall It would be great to get this merged whenever you get a chance. Then we can start working on things in more manageable chunks / PRs. 

I've had people messaging me asking about a dev release for 8.x. It would be best to get people to test this now that Message has a dev release on Drupal.org. Thanks!

RoySegall commented:
I saw you push more work. I'll have a look on this and if everything is OK i'll merge.

mccrodp commented:
@RoySegall - sounds good Roy. There will be some work remaining I'm sure, but if nothing major we can make some requirements toward a dev release after merge and I can create separate PRs against each one! 👯 What do you think?

RoySegall commented:
OK. I pushed the `8.x-1.x`. There are some [issues](https://github.com/RoySegall/message_ui/issues?q=is%3Aissue+is%3Aopen+label%3A%22Drupal+8%22) before i'll create an alpha release since this one relate to UX of the user.

Except that there are some coding standard issues which can be detect by the code sniffer. I'll add an issue and we can apply the same action as done in OG. 

God job!

mccrodp commented:
Thanks @RoySegall - that's fantastic. Yes there are certainly more issues to work on here, I'll create separate PRs now. Will you push a dev release to drupal.org in the meantime?

I have a few issues on Message that I want to work on now that this is merged also. So I'll ping you on those if I need some input.

Shall we close this one?

RoySegall commented:
No. Until the other issues we won't create a release.
